text/0000-unnamed-fields.md

```diff
@@ -0,0 +1,350 @@
+
+```rust
+struct S {
+    a: u32,
+    union {
+        b: u32,
+        struct {
+            c: u16,
+            d: f16,
+        },
+        e: f32,
+    },
+    f: u64,
+}
+```
+
+All of the following will instantiate a value of type `S`:
+
+- `S { a: 1, b: 2, f: 3.0 }`
+- `S { a: 1, c: 2, d: 3.0, f: 4.0 }`
+- `S { a: 1, e: 2.0, f: 3.0 }`
+
+## Representation
+
+By default, Rust lays out structures using its native representation,
+`repr(Rust)`; that representation permits any layout that can store all the
+non-overlapping fields simultaneously, and makes no other guarantees about the
+storage of unnamed fields.
+
+When using this mechanism to define a C interface, remember to use the
+`repr(C)` attribute to match C's data structure layout. Any representation
+attribute applied to the top-level structure also applies to every unnamed
+field within that declaration. Such a structure defined with `repr(C)` will use
+a representation identical to the same structure with all unnamed fields
+transformed to equivalent named fields of a struct or union type with the same
+fields.
+
+## Derive
+
+A `struct` or `union` containing unnamed fields may derive `Copy`, `Clone`, or
+both, if all the fields it contains (including within unnamed fields) also
+implement `Copy`.
+
+A `struct` containing unnamed fields may derive `Clone` if every field
+contained directly in the `struct` implements `Clone`, and every field
+contained within an unnamed `union` (directly or indirectly) implements `Copy`.
+
+## Ambiguous field names
+
+You cannot use this feature to define multiple fields with the same name. For
+instance, the following definition will produce an error:
+
+```rust
+struct S {
+    a: u32,
+    union {
+        a: u32,
+        b: f32,
+    },
+}
+```
+
+The error will identify the duplicate `a` fields as the sources of the error.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Parsing
+
+Within a struct's fields, in place of a field name and value, allow
+`union { fields }`, where `fields` allows everything allowed within a `union`
+declaration. Conversely, within a union's fields, in place of a field name
+and value, allow `struct { fields }`, where `fields` allows everything allowed
+within a `struct` declaration.
+
+Note that the keyword `struct` cannot appear as a field name, making it
+entirely unambiguous. The contextual keyword `union` could theoretically appear
+as a field name, but an open brace cannot appear immediately after a field
+name, allowing disambiguation via a single token of context (`union {`).
+
+## Layout and Alignment
+
+The layout and alignment of a `struct` or `union` containing unnamed
+fields should look the same as if each unnamed field has a separately declared
+type and a named field of that type, rather than as if the fields appeared
+directly within the containing `struct` or `union`. In some cases, this may
+result in different alignment.
+
```

Comment:
Rather than going into this much detail, suggest saying that the layout and alignment are whatever the C ABI requires for this case (because of `#[repr(C)]`), which is _probably_ the same as if the unnamed sub-structures were named.